README.md
<!--
 * @Description: In User Settings Edit
 * @Author: your name
 * @Date: 2019-04-08 15:51:04
 * @LastEditTime: 2019-08-28 12:17:49
 * @LastEditors: Please set LastEditors
 -->
# Python CLI Tool to Monitor Oracle Database

[TOC]

I name it **Pycle**, which means `Python` + `Oracle`. It's inspired by another open source python scripts - Pyora, and it's website is <https://github.com/bicofino/Pyora>

Pyora is good, but with some disadvantes, it does not support Python 3. So i try to use `Click`, `cx_Oracle` rewirte it, make it more compatible with Python 2 & Python 3.

Which `Click` is another great package to build CLI utility, official site: <https://click.palletsprojects.com/en/7.x/> 

> Click is a Python package for creating beautiful command line interfaces in a composable way with as little code as necessary. It’s the “Command Line Interface Creation Kit”. It’s highly configurable but comes with sensible defaults out of the box.
>
> It aims to make the process of writing command line tools quick and fun while also preventing any frustration caused by the inability to implement an intended CLI API.

Which cx_Oracle is developed by Oracle, official site: <https://cx-oracle.readthedocs.io/en/latest/>

> **cx_Oracle** is a module that enables access to Oracle Database and conforms to the Python database API specification. This module is currently tested against Oracle Client 11.2, 12.1, 12.2 and 18.3 and Python 2.7, 3.5, 3.6 and 3.7.

## How to install pycle

For easy to use, i built it as a python wheel package to resolve package dependency.

So you can download the wheel package and run bellow command to install on your machine

```shell
# which Pycle-0.0.2-py3-none-any.whl is on your local file system
$ pip install Pycle-0.0.2-py3-none-any.whl

# run test
$ pycle --help
Usage: pycle [OPTIONS] COMMAND [ARGS]...

  A CLI utility to check some database

  example:

  pycle chk-oracle --host='127.0.0.1' --port='1521', --sid='orcl'

Options:
  --help  Show this message and exit.

Commands:
  chk-oracle  查询Oracle
```

Or you can just download `pycle.py` scripts, but in this way you need to resolve dependency package by yourself

run bellow command

```shell
# install dependency package
$ pip install click
$ pip install cx_Oracle

# run test
$ python pycle.py --help           
Usage: pycle.py [OPTIONS] COMMAND [ARGS]...

  A CLI utility to check some database

  example:

  pycle chk-oracle --host='127.0.0.1' --port='1521', --sid='orcl'

Options:
  --help  Show this message and exit.

Commands:
  chk-oracle  查询Oracle
```

## How to Use

There are two way to use this utility, run `pycle` command or run `pycle.py` scripts directly. Before use to check Oracle database, you need create a user `pycle`, bellow is creation script

```shell
CREATE USER "PYCLE" IDENTIFIED BY "pycle";
grant "CONNECT" TO "PYCLE";
grant select on v_$instance to "PYCLE";
grant select on v_$session to "PYCLE";
grant select on v_$sysstat to "PYCLE";
grant select on v_$system_event to "PYCLE";
grant select on v_$asm_diskgroup to "PYCLE";
grant select on v_$sys_time_model to "PYCLE";
grant select on v_$parameter to "PYCLE";
grant select on v_$latch to "PYCLE";
grant select on v_$latch_children to "PYCLE";
grant select on v_$sgastat to "PYCLE";
grant select on v_$resource_limit to "PYCLE";
grant select on v_$datafile to "PYCLE";
grant select on v_$px_session to "PYCLE";
grant select on v_$dispatcher to "PYCLE";
grant select on v_$shared_server to "PYCLE";
grant select on v_$tablespace to "PYCLE";
grant select on v_$archived_log to "PYCLE";
grant select on v_$rman_status to "PYCLE";
grant select on v_$memory_dynamic_components to "PYCLE";
grant select on v_$log to "PYCLE";
grant select on v_$event_name to "PYCLE";
grant select on v_$archive_dest to "PYCLE";
grant select on v_$log_history to "PYCLE";
grant select on dba_alert_history to "PYCLE";
grant select on dba_objects to "PYCLE";
grant select on dba_users to "PYCLE";
grant select on dba_jobs_running to "PYCLE";
grant select on dba_sys_privs to "PYCLE";
grant select on dba_tab_privs to "PYCLE";
grant select on dba_tablespaces to "PYCLE";
grant select on dba_data_files to "PYCLE";
grant select on dba_free_space to "PYCLE";
grant select on dba_segments to "PYCLE";
```

- **Run pycle command**

  run pycle command would be easy and simple way, you don't need to give path to pycle.py to run. All options for chk-oracle command get a default value

  default values:

  ​	host = 127.0.0.1

  ​	port = 1521

  ​	sid = orcl

  ​	user = pycle

  ​	password = pycle

  ```shell
  # get help text
  $ pycle --help                                  
  Usage: pycle [OPTIONS] COMMAND [ARGS]...
  
    A CLI utility to check some database
  
    example:
  
    pycle chk-oracle --host='127.0.0.1' --port='1521', --sid='orcl'
  
  Options:
    --help  Show this message and exit.
  
  Commands:
    chk-oracle  查询Oracle
    
  # get help text for chk-oracle
  pycle chk-oracle --help
  echo first
  Usage: pycle chk-oracle [OPTIONS]
  
    查询Oracle
  
  Options:
    -h, --host TEXT                 数据库IP地址  [required]
    -p, --port TEXT                 数据库监听端口  [required]
    -s, --sid TEXT                  数据库实例SID  [required]
    -u, --user TEXT                 数据库用户名  [required]
    -P, --passwd TEXT               用户密码  [required]
    -c, --check [dbversion|dbsize|check_active|read_cache_hit_ratio|disk_sorts_ratio|count_of_active_users|size_of_user_data|dbfilesize|dbuptime|usercommits|userrollback|deadlocks|redowrites|tablescans|tablerowscans|indexffs|hardparseratio|netset|netrecv|netroundtrips|currentloggons|lastarch|lastapplyarch|freebuffwaits|busybuffwaits|logswitcompletion|logfilesync|logparallelwrite|enqueue|dbseqreadwait|dbscatteredread|dbsinglewrite|dbparallelwrite|directpathread|directpathwrite|latchfree|tablespace|tablespaceinuse]
                                    查询目标
    -C, --checkparam TEXT           查询参数
    --help                          Show this message and exit.
  ```

- **Run pycle.py script**

  To run pycle.py script is a little different from run pycle command

  ```shell
  $ python pycle.py --help
  Usage: pycle.py [OPTIONS] COMMAND [ARGS]...
  
    A CLI utility to check some database
  
    example:
  
    pycle chk-oracle --host='127.0.0.1' --port='1521', --sid='orcl'
  
  Options:
    --help  Show this message and exit.
  
  Commands:
    chk-oracle  查询Oracle
    
  $ python pycle/pycle.py chk-oracle --help
  echo first
  Usage: pycle.py chk-oracle [OPTIONS]
  
    查询Oracle
  
  Options:
    -h, --host TEXT                 数据库IP地址  [required]
    -p, --port TEXT                 数据库监听端口  [required]
    -s, --sid TEXT                  数据库实例SID  [required]
    -u, --user TEXT                 数据库用户名  [required]
    -P, --passwd TEXT               用户密码  [required]
    -c, --check [dbversion|dbsize|check_active|read_cache_hit_ratio|disk_sorts_ratio|count_of_active_users|size_of_user_data|dbfilesize|dbuptime|usercommits|userrollback|deadlocks|redowrites|tablescans|tablerowscans|indexffs|hardparseratio|netset|netrecv|netroundtrips|currentloggons|lastarch|lastapplyarch|freebuffwaits|busybuffwaits|logswitcompletion|logfilesync|logparallelwrite|enqueue|dbseqreadwait|dbscatteredread|dbsinglewrite|dbparallelwrite|directpathread|directpathwrite|latchfree|tablespace|tablespaceinuse]
                                    查询目标
    -C, --checkparam TEXT           查询参数
    --help                          Show this message and exit.
  ```

- Example

  Say i want to get a database logical size, run command like bellow

  ```shell
  $ pycle chk-oracle --host=172.16.1.204 --sid=shs --check=dbsize
  13667270656
  ```

## Dev Environments Configuration

Say you try to build your scripts, here is a brief introduction for how to get your environment done and build your CLI utility.

- Create Virtual env

  ```shell
  $ virtualenv --no-site-packages env
  ```

- `Click` installation

  ```shell
  $ pip install click
  ```

- `cx_Oracle` installation

  ```shell
  $ python -m pip install cx_Oracle --upgrade
  ```

  - Client side configuration

    After install cx_Oracle, package,  you need to configure correctly your environment to use cx_Oracle. I'm using macOS, so check bellow content for how to configure client on macOS. for other platform, such as Windows, Linux, choose specific installation guide on this [url](https://cx-oracle.readthedocs.io/en/latest/installation.html#).

    1. Download Oracle instant client from [here](http://www.oracle.com/technetwork/topics/intel-macsoft-096467.html). Choose either a 64-bit or 32-bit package, matching your Python architecture.

    2. Unzip the package into a single directory that is accessible to your application. For example:

       ```shell
       mkdir -p /opt/oracle
       unzip instantclient-basic-macos.x64-12.2.0.1.0.zip
       ```

    3. Add links to `$HOME/lib` or `/usr/local/lib` to enable applications to find the library. For example:

       ```shell
       mkdir ~/lib
       ln -s /opt/oracle/instantclient_12_2/libclntsh.dylib ~/lib/
       ```

       Alternatively, copy the required OCI libraries. For example:

       ```shell
       mkdir ~/lib
       cp /opt/oracle/instantclient_12_2/{libclntsh.dylib.12.1,libclntshcore.dylib.12.1,libons.dylib,libnnz12.dylib,libociei.dylib} ~/lib/
       ```

       For Instant Client 11.2, the OCI libraries must be copied. For example:

       ```shell
       mkdir ~/lib
       cp /opt/oracle/instantclient_11_2/{libclntsh.dylib.11.1,libnnz11.dylib,libociei.dylib} ~/lib/
       ```

    4. If you intend to co-locate optional Oracle configuration files such as `tnsnames.ora`, `sqlnet.ora`or `oraaccess.xml` with Instant Client, then create a `network/admin` subdirectory. For example:

       ```shell
       mkdir -p /opt/oracle/instantclient_12_2/network/admin
       ```

       This is the default Oracle configuration directory for executables linked with this Instant Client.

       Alternatively, Oracle configuration files can be put in another, accessible directory. Then set the environment variable `TNS_ADMIN` to that directory name.

### How to Pack

If you wan to make your distribution package, you can check this instructions to learn how

Setup tools: <https://click.palletsprojects.com/en/7.x/setuptools/>

Python Packaging User Guide: https://packaging.python.org/#python-packaging-user-guide

Changelog.md
<!--
 * @Author: John Wong
 * @Date: 2019-04-08 16:00:03
 * @LastEditors: John Wong
 * @LastEditTime: 2019-08-28 12:48:50
 * @Description: file content
 -->
- add select on v$log_history privilage to user pycle on README.md doc
- correct some sql statments

pycle/pycle.py
#!/usr/bin/env python
# coding=UTF-8
'''
@Author: John Wong
@LastEditors: John Wong
@Description: 
@Date: 2019-03-29 09:18:05
@LastEditTime: 2019-08-28 12:46:19
'''

import click
import cx_Oracle

# sql 查询语句
sqls = {
    "dbversion": '''
                 select banner from v$version where rownum=1
                 ''',
    "dbsize": '''
              select sum(bytes) from dba_segments
              ''',
    "check_active": '''
                    select to_char(case when inst_cnt > 0 then 1 else 0 end,
                    'fm99999999999999990') retvalue from (select count(*) inst_cnt
                    from v$instance where status = 'open' and logins = 'allowed'
                    and database_status = 'active')
                    ''',
    "read_cache_hit_ratio": '''
                            select nvl(to_char((1 - (phy.value - lob.value - dir.value) /
                            ses.value) * 100, 'fm99999990.9999'), '0') retvalue
                            from   v$sysstat ses, v$sysstat lob,
                            v$sysstat dir, v$sysstat phy
                            where  ses.name = 'session logical reads'
                            and    dir.name = 'physical reads direct'
                            and    lob.name = 'physical reads direct (lob)'
                            and    phy.name = 'physical reads'
                            ''',
    "disk_sorts_ratio": '''
                        select nvl(to_char(d.value/(d.value + m.value)*100,
                        'fm99999990.9999'), '0') retvalue
                        from  v$sysstat m, v$sysstat d
                        where m.name = 'sorts (memory)'
                        and d.name = 'sorts (disk)'
                        ''',
    "count_of_active_users": '''
                             select to_char(count(*), 'fm99999999999999990') retvalue
                             from v$session where username is not null
                             and status='ACTIVE'
                             ''',
    "size_of_user_data": '''
                        select to_char(sum(  nvl(a.bytes - nvl(f.bytes, 0), 0)),
                        'fm99999999999999990') retvalue
                        from sys.dba_tablespaces d,
                        (select tablespace_name, sum(bytes) bytes from dba_data_files
                        group by tablespace_name) a,
                        (select tablespace_name, sum(bytes) bytes from
                        dba_free_space group by tablespace_name) f
                        where d.tablespace_name = a.tablespace_name(+) and
                        d.tablespace_name = f.tablespace_name(+)
                        and not (d.extent_management like 'local' and d.contents
                        like 'temporary')
                        ''',
    "dbfilesize": '''
                  select to_char(sum(bytes), 'fm99999999999999990') retvalue
                  from dba_data_files
                  ''',
    "dbuptime": '''
                select to_char((sysdate-startup_time)*86400,
                'fm99999999999999990') retvalue from v$instance
                ''',
    "usercommits": '''
                   select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
                   v$sysstat where name = 'user commits'
                   ''',
    "userrollback": '''
                    select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
                    v$sysstat where name = 'user rollbacks'
                    ''',
    "deadlocks": '''
                 select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
                 v$sysstat where name = 'enqueue deadlocks'
                 ''',
    "redowrites": '''
                  select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
                  v$sysstat where name = 'redo writes'
                  ''',
    "tablescans": '''
                  select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
                  v$sysstat where name = 'table scans (long tables)'
                  ''',
    "tablerowscans": '''
                     select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
                     v$sysstat where name = 'table scan rows gotten'
                     ''',
    "indexffs": '''
                select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
                v$sysstat where name = 'index fast full scans (full)'
                ''',
    "hardparseratio": '''
                      select nvl(to_char(h.value/t.value*100,'fm99999990.9999'), '0')
                      retvalue from  v$sysstat h, v$sysstat t where h.name = 'parse count (hard)'
                      and t.name = 'parse count (total)'
                      ''',
    "netsent": '''
              select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
              v$sysstat where name = 'bytes sent via SQL*Net to client'
              ''',
    "netrecv": '''
               select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
               v$sysstat where name = 'bytes received via SQL*Net from client'
               ''',
    "netroundtrips": '''
                     select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
                     v$sysstat where name = 'SQL*Net roundtrips to/from client'
                     ''',
    "currentloggons": '''
                      select nvl(to_char(value, 'fm99999999999999990'), '0') retvalue from
                      v$sysstat where name = 'logons current'
                      ''',
    "lastarch": '''
                select to_char(max(sequence#), 'fm99999999999999990')
                retvalue from v$log where archived = 'YES'
                ''',
    "lastapplyarch": '''
                     select to_char(max(lh.sequence#), 'fm99999999999999990')
                     retvalue from v$log_history lh, v$archived_log al
                     where lh.sequence# = al.sequence# and applied='YES'
                     ''',
    "freebuffwaits": '''
                     select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                     from v$system_event se, v$event_name en
                     where se.event(+) = en.name and en.name = 'free buffer waits'
                     ''',
    "busybuffwaits": '''
                     select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                     from v$system_event se, v$event_name en where se.event(+) =
                     en.name and en.name = 'buffer busy waits'
                     ''',
    "logswitcompletion": '''
                         select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                         from v$system_event se, v$event_name en where se.event(+)
                         = en.name and en.name = 'log file switch completion'
                         ''',
    "logfilesync": '''
                   select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                   from v$system_event se, v$event_name en
                   where se.event(+) = en.name and en.name = 'log file sync'
                   ''',
    "logparallelwrite": '''
                        select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                        from v$system_event se, v$event_name en where se.event(+)
                        = en.name and en.name = 'log file parallel write'
                        ''',
    "dbseqreadwait": '''
                     select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                     from v$system_event se, v$event_name en where se.event(+)
                     = en.name and en.name = 'db file sequential read'
                     ''',
    "dbscatteredread": '''
                        select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                        from v$system_event se, v$event_name en where se.event(+)
                        = en.name and en.name = 'db file scattered read'
                        ''',
    "dbsinglewrite": '''
                        select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                        from v$system_event se, v$event_name en where se.event(+)
                        = en.name and en.name = 'db file single write'
                     ''',
    "dbparallelwrite": '''
                        select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                        from v$system_event se, v$event_name en where se.event(+)
                        = en.name and en.name = 'db file parallel write'
                        ''',
    "directpathread": '''
                        select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                        from v$system_event se, v$event_name en where se.event(+)
                        = en.name and en.name = 'direct path read'
                        ''',
    "directpathwrite": '''
                        select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                        from v$system_event se, v$event_name en where se.event(+)
                        = en.name and en.name = 'direct path write'
                       ''',
    "latchfree": '''
                 select nvl(to_char(time_waited, 'fm99999999999999990'), '0') retvalue
                 from v$system_event se, v$event_name en where se.event(+)
                 = en.name and en.name = 'latch free'
                 ''',
    "tablespace": '''
                    select  tablespace_name,
                    100-(trunc((max_free_mb/max_size_mb) * 100)) as used
                    from ( select a.tablespace_name,b.size_mb,a.free_mb,b.max_size_mb,a.free_mb + (b.max_size_mb - b.size_mb) as max_free_mb
                    from   (select tablespace_name,trunc(sum(bytes)/1024/1024) as free_mb from dba_free_space group by tablespace_name) a,
                    (select tablespace_name,trunc(sum(bytes)/1024/1024) as size_mb,trunc(sum(greatest(bytes,maxbytes))/1024/1024) as max_size_mb
                    from   dba_data_files group by tablespace_name) b where  a.tablespace_name = b.tablespace_name
                    ) where tablespace_name='{0}' order by 1
                  ''',
    "tablespaceinuse": '''
                        select df.tablespace_name "tablespace", (df.totalspace -
                        tu.totalusedspace) "freemb" from (select tablespace_name,
                        sum(bytes) totalspace from dba_data_files group by tablespace_name)
                        df ,(select sum(bytes) totalusedspace,tablespace_name from dba_segments
                        group by tablespace_name) tu where tu.tablespace_name =
                        df.tablespace_name and df.tablespace_name = '{0}'
                        ''',
}

check_item = tuple(sqls.keys())


# @click.group(chain=True)
@click.group()
def main():
    '''
    A CLI utility to check some database \n
    example: \n
    pycle chk-oracle --host='127.0.0.1' --port='1521', --sid='orcl'
    '''
    pass

# @click.group()
@main.command(help=u'查询Oracle')
@click.option('-h', '--host', default='127.0.0.1', required=True, help=u'数据库IP地址')
@click.option('-p', '--port', default='1521', required=True, help=u'数据库监听端口')
@click.option('-s', '--sid', default='orcl', required=True, help=u'数据库实例SID')
@click.option('-u', '--user', default='PYCLE', required=True, help=u'数据库用户名')
@click.option('-P', '--passwd', default='pycle', required=True, help=u'用户密码')
@click.option('-c', '--check', default='dbversion', required=False, type=click.Choice(check_item), help='查询目标')
@click.option('-C', '--checkparam', default='', required=False, multiple=True, help=u'查询参数')
def chk_oracle(host, port, sid, user, passwd, **kwargs):
    '''
    查询Oracle
    '''
    # get check type, default is dbversion
    check = kwargs.get('check')
    # get sql statement
    sql = sqls.get(check).strip()
    # conn to oracle
    dsn = cx_Oracle.makedsn(host=host, port=port, sid=sid)
    try:
        connection = cx_Oracle.connect(
            user=user, password=passwd, dsn=dsn, encoding='utf-8')
        cursor = connection.cursor()
        cursor.execute(sql)
        ret = cursor.fetchall()
        for i in ret:
            # 如果返回空值，那么额外处理
            click.echo(i[0])
        connection.close()
    except Exception as e:
        click.echo(str(e))


# @main.command(help=u'查询Mysql --待开发')
# def chk_mysql():
#     pass


# @main.command(help=u'查询mongo --待开发')
# def chk_mongo():
#     pass


if __name__ == "__main__":
    main()
